### DISCUSSION TITLE
Saving and Loading while Retaining Tags

### DISCUSSION THREAD

**Discussion by oiuytrewqp** [inkle/ink#784] [Game integration help] [answered]
I just finished an initial version of my game called My Isekai Dream. I added a 'Show and Tell' discussion: https://github.com/inkle/ink/discussions/783

I use tags to update backgrounds and displayed characters. Most of the tags are set once and persist until changed. e.g. the background only needs to be set once and then will remain until it is changed.

However when I save and then load I lose the history of the tags, as they where set on previous text.

Has any one used tags like this? Any suggestions overcoming the issue?

For example in the extracted sample if I save on the last line of text and then load, the tag 'background', which is on the first line is not available.

```
=== school ===
# background: School_Hallway_Day
# portrait: Keiko.girl.normal.seifuku2.longdark.longdark
# characters: Keiko.none
"Ok. Bye."\\nKeiko waves and runs off to meet her girlfriends. They have been waiting for her. They all immediately break into chatter.
# portrait: Hiroshi.boy.normal.uniform.short3brown.shortbrown
# characters: Hiroshi.none
"I'm going to meet Koshiro."\\nHiroshi gives a halfhearted wave and walks off.
# portrait: Ryoichi.boy.normal.uniform.short2silver.short2silver
"Just us two, eh?" Ryoichi says, "Want to head to class... and on the way, we can scout who wrote that letter..."
```

Thanks

**Comment by oiuytrewqp** [comment, marked as answer]
Never mind. I've found out that I have misunderstood tags; they are transient and only affect the current line.

What Ill do to allow saving of persistent updates is to make use of global variables. e.g. VAR background = ...

Global variables are saved as part of the current state and will fix my issue.

**Reply by pookage** [reply]
I'm just starting out and wondering how to save my state - is save-game functionality built-in, or is that something I need to make myself separately?

**Reply by oiuytrewqp** [reply]
Wow, thats going back. I used c# in Godot.

Looking at the code Ink provides the save data, but I had to save and load the files myself.

To load, what I did was create a Story story = new Story(ink_json), where ink_json is the full ink file. This loads the ink file only. After that I needed to set the state which was saved, by calling story.state.LoadJson(save_json), where the save_json is the saved state.

To Save, I just had to get the latest state by calling story.sate.ToJson().

Summary:

Setup Ink story before load and save: Story story = new Story(ink_json)

Save a state: story.sate.ToJson()

Load a state: story.state.LoadJson(save_json)

Handle the json saving yourself, dependent on platform, etc.

**Reply by pookage** [reply]
Incredible! Thank you for looking back and helping an internet stranger! 🙏
